### PR TITLE
chore: rebalancer review

### PR DESCRIPTION
**Pull request by Picodes**

### CODE REVIEW

File: contracts/helpers/Rebalancer.sol [hunk 1]
```diff
@@ -52,16 +53,12 @@ contract Rebalancer is AccessControl {
...
-    /// @param amountOutMin Minimum amount of `tokenOut` that must be obtained from the swap
-    /// @param to Address to which `tokenOut` must be sent
-    /// @param deadline Timestamp before which this transaction must be included
-    /// @return amountOut Amount of outToken obtained
+    /// @inheritdoc IRebalancer
     /// @dev Contrarily to what is done in the Transmuter contract, here neither of `tokenIn` or `tokenOut`
     /// should be an `agToken`
+    /// @dev Can be used even if the subsidy budget is 0, in which case it'll just do 2 transmuter swaps
+    /// @dev The invariants should be that `msg.sender` injects `amountIn` in the transmuter and either the
+    /// subsidy is 0 either he receives the guaranteed amount out from this contract + the transmuter
```
Comment (sogipec) [inline contracts/helpers/Rebalancer.sol:61]: either they receive the guaranteed amount out

File: contracts/interfaces/IRebalancer.sol [hunk 2]
```diff
@@ -0,0 +1,68 @@
+// SPDX-License-Identifier: GPL-3.0
+
+pragma solidity ^0.8.19;
```
Comment (sogipec) [inline contracts/interfaces/IRebalancer.sol:3]: for interfaces, let's do pragma solidity >= 0.5.0;

File: contracts/interfaces/IRebalancer.sol [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+
+import "oz/interfaces/IERC20.sol";
+import "oz/interfaces/IERC20Metadata.sol";
+import "oz/token/ERC20/utils/SafeERC20.sol";
+
+import { ITransmuter } from "interfaces/ITransmuter.sol";
+
+import { AccessControl, IAccessControlManager } from "../utils/AccessControl.sol";
```
Comment (sogipec) [inline contracts/interfaces/IRebalancer.sol:11]: remove useless imports here, and import them from the base file

File: contracts/helpers/Rebalancer.sol [hunk 4]
```diff
@@ -158,30 +161,36 @@ contract Rebalancer is AccessControl {
...
     /// @dev Before calling this function, governance must make sure that there are enough `agToken` idle
     /// in the contract to sponsor the swaps
+    /// @dev This function can be used to decrease an order by overriding it
     function setOrder(
         address tokenIn,
         address tokenOut,
         uint256 subsidyBudget,
         uint256 guaranteedRate
     ) external onlyGuardian {
+        uint8 decimalsIn = TRANSMUTER.getCollateralDecimals(tokenIn);
```
Comment (GuillaumeNervoXS) [inline contracts/helpers/Rebalancer.sol:174]: Can we only call them only if it wasn't set already in a previous order?

File: contracts/helpers/Rebalancer.sol [hunk 5]
```diff
@@ -158,30 +161,36 @@ contract Rebalancer is AccessControl {
...
         uint256 newBudget = budget + subsidyBudget - order.subsidyBudget;
-        if (IERC20(agToken).balanceOf(address(this)) < newBudget) revert InvalidParam();
+        if (IERC20(AGTOKEN).balanceOf(address(this)) < newBudget) revert InvalidParam();
         budget = newBudget;
-        order.subsidyBudget = subsidyBudget;
-        order.guaranteedRate = guaranteedRate;
+        order.subsidyBudget = subsidyBudget.toUint112();
+        order.decimalsIn = decimalsIn;
+        order.decimalsOut = decimalsOut;
+        order.guaranteedRate = guaranteedRate.toUint128();
```
Comment (GuillaumeNervoXS) [inline contracts/helpers/Rebalancer.sol:185]: why not make the params passed directly as `uint128 guaranteedRate``

File: contracts/helpers/Rebalancer.sol [hunk 6]
```diff
@@ -158,30 +161,36 @@ contract Rebalancer is AccessControl {
...
-            revert NotCollateral();
+        if (decimalsIn == 0 || decimalsOut == 0) revert NotCollateral();
         Order storage order = orders[tokenIn][tokenOut];
         uint256 newBudget = budget + subsidyBudget - order.subsidyBudget;
-        if (IERC20(agToken).balanceOf(address(this)) < newBudget) revert InvalidParam();
+        if (IERC20(AGTOKEN).balanceOf(address(this)) < newBudget) revert InvalidParam();
         budget = newBudget;
-        order.subsidyBudget = subsidyBudget;
-        order.guaranteedRate = guaranteedRate;
+        order.subsidyBudget = subsidyBudget.toUint112();
```
Comment (GuillaumeNervoXS) [inline contracts/helpers/Rebalancer.sol:182]: why not make the params passed directly as `uint112 subsidyBudget`

**Review by GuillaumeNervoXS — APPROVED** [review]

**Pull request merged**